src/category_theory/limits/shapes/types.lean

```diff
@@ -0,0 +1,107 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.limits.types
+import category_theory.limits.shapes.products
+import category_theory.limits.shapes.binary_products
+import category_theory.limits.shapes.terminal
+
+/-!
+# Special shapes for limits in `Type`.
+
+The general shape (co)limits defined in `category_theory.limits.types`
+are intended for use through the limits API,
+and the actual implementation should mostly be considered "sealed".
+
+In this file, we provide definitions of the "standard" special shapes of limits in `Type`,
+giving the expected definitional implementation:
+* the terminal object is `punit`
+* the binary product of `X` and `Y` is `X × Y`
+* the product of a family `f : J → Type` is `Π j, f j`.
+
+It is not intended that these definitions will be global instances:
+they should be turned on as needed.
+
+The exception to this rule is that the monoidal category structure on `Type`
```

Comment:
Why will this not cause a lot of pain down the road?

Reply:
I think we can dig ourselves back out of this hole fairly cheaply. Try out the following code at the end of this file:

```
def compare_limits {J : Type v} [small_category J] {C : Type u} [category.{v} C] (F : J ⥤ C)
  {h₁ : has_limit F} {h₂ : has_limit F} :
  (by { haveI := h₁, exact limit F }) ≅ (by { haveI := h₂, exact limit F }) :=
is_limit.cone_point_unique_up_to_iso (h₁.is_limit) (h₂.is_limit)

example (X Y : Type) : limits.prod X Y ≅ X ⊗ Y :=
iso.refl _ -- fails with `synthesized type class instance is not definitionally equal to expression inferred by typing rules`

example (X Y : Type) : limits.prod X Y ≅ X ⊗ Y :=
compare_limits _ -- works fine
```

`iso.refl _` fails, because there are two completely unrelated definitions of the limit: the "general" one is being pulled in for the `limits.prod`, but the "nice" one is baked into the monoidal structure and used in `X ⊗ Y`.

However `compare_limits` can extract these `has_limit` instances back out of the implicit arguments, and use them to construct an isomorphism comparing the two results.

`compare_limits` should be added to mathlib; I'll get on it.

Reply:
I guess you would also need `compare_products` and `compare_equalizers` etc???

Reply:
But I agree that this would solve the issue

Reply:
Notice here that the generic `compare_limits` happily constructed the isomorphism for products. No need for special shapes, I think.